事業をエンジニアリングしよう ②実装

u-toru42 opened the pull request:
RSpecの結果です。
![スクリーンショット 2022-11-20 20 32 19](https://user-images.githubusercontent.com/37538568/202899685-de72ad28-a9fd-4e3e-b27c-2e0a7cca87c3.png)

プロフィール画面にhobby, profileの欄を追加しました。
![スクリーンショット 2022-11-16 20 35 04](https://user-images.githubusercontent.com/37538568/202170269-f01cc85d-d823-419d-8fba-aad416e68170.png)
イベント画面から遷移できるプロフィール画面です。フォローボタンに合わせてフォロワー数が増減します。
![スクリーンショット 2022-11-20 20 33 42](https://user-images.githubusercontent.com/37538568/202899741-3ef1247c-f3e9-4e9b-b1cb-3f26af0c206d.png)

マイページに自分のフォロー/フォロワーのリンクを貼りました。
![スクリーンショット 2022-11-20 20 34 27](https://user-images.githubusercontent.com/37538568/202899763-32ca2177-ec5e-45c3-99ff-2bb8fc6e1beb.png)

フォロー一覧画面です。
![スクリーンショット 2022-11-16 20 33 14](https://user-images.githubusercontent.com/37538568/202169883-db76f163-17cd-4125-b786-396c21777899.png)
フォロワー一覧画面です。
![スクリーンショット 2022-11-16 20 33 52](https://user-images.githubusercontent.com/37538568/202170017-569f81f8-ad98-4573-9613-4280f386fe52.png)

u-toru42 committed:
環境構築でgem追加

u-toru42 committed:
課題用のSpecのzipファイル

u-toru42 committed:
プロフィール表示機能の追加

u-toru42 committed:
params.permitにhobbyとprofileを追加

u-toru42 committed:
プロフィールリンクを追加し、ビューを調整

u-toru42 committed:
webpackの構成を追加

u-toru42 committed:
profileを読み込みできるようにする

u-toru42 committed:
Specの追加

u-toru42 committed:
hobbyとprofileのカラム追加

u-toru42 committed:
hobbyとprofileのカラムを追加

u-toru42 committed:
課題用Specの追加

u-toru42 committed:
Railsチュートリアルを参考に構成(途中)

u-toru42 committed:
profile間でfollow関係をまとめる

u-toru42 committed:
followメソッドとunfollowメソッドを紐付けて扱う

u-toru42 committed:
follow機能とプロファイル機能を実装

u-toru42 committed:
バリデーションを削除

u-toru42 committed:
Railsチュートリアルの実装方法から変更する

u-toru42 committed:
参加者のプロファイル機能を実装

u-toru42 committed:
イベント主催者のプロファイル機能を実装

u-toru42 committed:
詳細画面の主催者、参加者のnameにprofile画面に遷移するリンクを実装

u-toru42 committed:
プロフィール画面の調整

u-toru42 committed:
ヘッダーのフォロー/フォロワー画面を削除

u-toru42 committed:
フォロー/フォロワーの一覧画面を実装

u-toru42 committed:
users範囲内にfollow機能をまとめる

u-toru42 committed:
微調整

u-toru42 committed:
フォロー関係のカラム名と内容を修正

u-toru42 committed:
コメントを修正

u-toru42 committed:
Profile経由のfollowパーシャルとフォロー/フォロワー数の計測を削除

u-toru42 committed:
余白の削除

u-toru42 committed:
遷移先のidの反映バグ修正

u-toru42 committed:
遷移先のidバグ修正

u-toru42 committed:
マイページ画面で自分のフォロー/フォロワー画面を見れる

u-toru42 committed:
遷移先の修正とフォロー関係の表示の追加

u-toru42 committed:
一覧で不要な項目を削除

u-toru42 committed:
修正

u-toru42 committed:
修正

Tsuchiya2 commented on app/controllers/profiles_controller.rb:
```diff
@@ -0,0 +1,28 @@
+class ProfilesController < ApplicationController
+  def show
+    # プロファイル全般
+    @user = User.find_by(id: params[:id])
```
idを探し出す際はfind_byではなく、もっと別なメソッドの方が良いですね。

u-toru42 replied:
ご指摘頂きましたので、find_byからメソッドをfindに変え、メモ化してDB負荷を下げました。
3140a24198dcf989f11ea63eb47dc48f62b7bd63>こちらで対応を行いました。

Tsuchiya2 commented on app/decorators/profile_decorator.rb:
```diff
@@ -0,0 +1,13 @@
+class ProfileDecorator < Draper::Decorator
+  delegate_all
```
使っていないようでしたら削除しておきましょう。（他の箇所も同様です）

u-toru42 replied:
ご指摘頂きましたので、profileとrelationshipのdecorator.rbを削除しました。
a242afbf394bfb1bd0999a1725ecb41bbe8327ae>こちらで対応を行いました。

Tsuchiya2 commented on app/views/events/_attendee.html.erb:
```diff
@@ -1,6 +1,6 @@
   <div class="d-flex align-items-center gap-2 mb-3">
     <%= image_tag attendee.decorate.avatar, class: 'rounded-circle avatar-md' %>
-    <%= attendee.name %>
+    <%= link_to attendee.name, profile_path(attendee.id) %>
```
`attendee.id`と記載するのはあまりしないですね。
オブジェクトを渡しても同じ結果になると思うので修正しましょう。（他の箇所も同様です）

u-toru42 replied:
ご指摘頂きましたので、attendee, event, @userのidの記載を削除しました。
3e0b5f959ae4599ac3572ca449ff389953b83793>こちらで対応を行いました。

Tsuchiya2 commented on app/views/shared/_header.html.erb:
```diff
@@ -11,6 +11,7 @@
       <% end %>
       <% if logged_in? %>
         <ul class="navbar-nav ms-auto mb-2 mb-lg-0">
+          </li>
```
削除のし忘れですかね？

u-toru42 replied:
ご指摘頂きましたので、削除のし忘れを修正しました。
3de0487082370a28923208f383e74e07b1a918ba>こちらで対応を行いました。

Tsuchiya2 commented on app/views/users/followers.html.erb:
```diff
@@ -0,0 +1,20 @@
+      <h1 class="font-weight-normal mt-5 mb-5">フォロワー一覧</h1>
+    </div>
+    <div class="row justify-content-center">
+      <% @users.each do |user| %>
```
eachで回して表示ではなく、collectionを使った方がパフォーマンスが良さそうです。（他の箇所も同様です）

u-toru42 replied:
ご指摘頂きましたので、collectionで実装しました。
4c80e84f43b4ae0e9f89843133ad14325442e3ad>こちらで対応を行いました。

Tsuchiya2 commented on app/controllers/profiles_controller.rb:
```diff
@@ -0,0 +1,28 @@
+    @posts = @user.posts.page(params[:page]).reverse_order
+    @following_users = @user.following_user
+    @follower_users = @user.follower_user
+  end
```
どのアクションでもユーザーレコードを呼び出しているので共通化しましょう。
また、@following_users, @follower_userも共通化できそうです。（似たような他のファイルも同様です）

u-toru42 replied:
ご指摘頂きましたので、処理を共通化し、メモ化しました。また、@following_users, @follower_userを共通化しました。
3140a24198dcf989f11ea63eb47dc48f62b7bd63>こちらで対応を行いました。

Tsuchiya2 reviewed (commented):
一旦気になる部分をコメントしました。

u-toru42 committed:
処理を共通化、メモ化

・ユーザーレコードの呼び出しを共通化し、メモ化してDB負荷を下げました。
・before_actionにユーザーレコードの呼び出しを入れ、@following_users, @follower_userを共通化しました。

u-toru42 committed:
profile, relationshipのDraperは使わないので削除します

u-toru42 committed:
idの記載を修正しました。オブジェクトを渡して処理します。

u-toru42 committed:
消し忘れを修正しました。

u-toru42 committed:
collectionで実装

eachの処理をcollectionを使うように修正しました。また、パーシャルにして共通化しました。

u-toru42 committed:
修正、削除

busitora commented on app/views/users/followers.html.erb:
```diff
@@ -0,0 +1,12 @@
+<div class="cover-user-index">
```
app/views/users/follows.html.erb と同じであれば共通化できますね

u-toru42 replied:
ご指摘頂きましたので、follows、followersのビューを render partialを使い、共通化しました。
d925c51a60a8d64a99c6c5f4bd325dced98bc66c>こちらで対応を行いました。

busitora commented on app/controllers/concerns/profile_follow.rb:
```diff
@@ -0,0 +1,23 @@
+require 'active_support'
```
インスタンス変数を使わないように実装できますか？

u-toru42 replied:
ご指摘頂きましたので、インスタンス変数の@following_users, @follower_userを使わずに、@user、@usersで管理できるように実装致しましたが、このような理解で合っておりますか。
b48381b70038021c029e7f6b851a7744f27088f8>こちらで対応を行いました。

busitora replied:
app/controllers/concerns/follow_profile.rb に対してですね！

controllerで使う分には問題ないです！

u-toru42 replied:
お世話になっております。遅まきながら、ご指示の内容を理解できた気がします。
これまでconcerns内でコントローラの処理をまとめる流れの解釈が浅く、変数なしでユーザーレコードの呼び出しの処理のみを実装するという考えに至れず、ご迷惑をおかけしました。
下記のような実装を致しましたが、格段に見通しが良くなり、管理もしやすくなった気がします。
f02b11c075e7deba42a3b1a6e1609909f0ccfef1>こちらで対応を行いました。

u-toru42 committed:
フォロー関係のビューを共通化

・フォロー/フォロワーの一覧の表示判別はif文を使い、pathで分岐
・collection機能は維持

u-toru42 committed:
before_actionにonlyオプションを実装

・サインアップやログインでbefore_actionが実行されると不具合が生じるため

u-toru42 committed:
rubocopによるコード修正

・主に# frozen_string_literal: trueと、""を''に変換

u-toru42 committed:
rubocop追加

u-toru42 committed:
データベースの構成を修正

・followed->followingに名前を修正
・followのDBはreferenceを使わないように変更

u-toru42 committed:
インスタンス変数を使わない実装

・@following_users, @follower_userを使わずに実装
・ProfileFollowは英語の文脈が正しくないのでFollowProfileにモジュール名を修正

u-toru42 committed:
path名に合わせてビューを修正

・フォロー、フォロワー数を数を前に設置し、リンク範囲を広げる
・マイページも同様の見た目とする

u-toru42 committed:
マイページのフォロー/フォロワーリンクの配置を横並びに修正

u-toru42 committed:
モデル内でインスタンス変数を使わずに処理を実装

・follow_profile.rbのメソッドのインスタンス変数を削除し、ユーザーレコードの呼び出しの処理のみ実装
・users_controller、profile_controllerには@user、@usersにそれぞれ処理を代入するのみとする

busitora commented:
LGTM！

u-toru42 commented:
> LGTM！

ありがとうございます！
